openedx/core/djangoapps/discussions/api/external_discussions.py

```diff
@@ -0,0 +1,73 @@
+from django.db import IntegrityError
+from opaque_keys.edx.keys import LearningContextKey, UsageKey
+
+from .data import ExternalDiscussionData
+from ..models import ExternalDiscussionsIdMapping
+
+
+def create_external_discussion_mapping(
+    context_key: LearningContextKey,
+    usage_key: UsageKey,
+    external_discussion_id: str
+) -> ExternalDiscussionData:
+    """
+    Create an ExternalDiscussionsIdMapping given the context key, usage key, and external discussion id
+    """
+
+    try:
+        discussion_mapping = ExternalDiscussionsIdMapping.objects \
+            .create(
+                context_key=context_key,
+                usage_key=usage_key,
+                external_discussion_id=external_discussion_id
+            )
+    except IntegrityError:
+        raise ValueError(
```

Comment:
Any particular reason we chose to re-raise this as a `ValueError`? It seems less informative than `IntegrityError` here:

From the docs [1]:

> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

> [...] but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a ValueError.

I can imagine some cases (_technically_, it is an error with the value, albeit conditionally based on the state of the database), but curious to your thinking.

Semantics aside, I'd be worried about code possibly catching the "wrong" `ValueError` since it's a system builtin and could be thrown from other lines of code.

- [1] https://docs.python.org/3/library/exceptions.html#ValueError

Reply:
@stvstnfrd 

I am not opposed to using `IntegrityError` here if we would like. I was following [PR comment](https://github.com/edx/edx-platform/pull/24565#discussion_r459539124) above when using `ValueError`. Please let me know how you'd like to proceed.

> Could you check for exceptions here? One might come up if violate uniqueness. That should be captured here and we should return a ValueError (I think that makes the most sense?)

Reply:
@stvstnfrd I recommended this to isolate internal errors from plugins. It's possible a future change causes this to no longer raise an IntegrityError, but still be invalid. 

To avoid the situation you describe, perhaps we can subclass ValueError?